Add Hashing and Equality typeclasses.

axel22 opened the pull request:
Modify TrieMap to use hashing and equality.

Modify serialization in TrieMap appropriately.

Review by @jsuereth @retronym

Commit:
Add Hashing and Equality typeclasses.

Modify TrieMap to use hashing and equality.

Modify serialization in TrieMap appropriately.

axel22 commented:
Review by @jsuereth

retronym commented on src/library/scala/Equality.scala:
```diff
@@ -0,0 +1,37 @@
+  * @since 2.10
+  */
+@annotation.implicitNotFound(msg = "No implicit Equality defined for ${T}.")
+trait Equality[T] extends Serializable {
```
We've already got `scala.math.Equiv`.

axel22 replied:
True - we should use that. We should also make that one Serializable.

retronym replied:
This is a pretty tricky area: you can't make these guys contravariant, as Paul's thread highlights; it's not clear how to associate compatible Eq/Hash; type class based equality doesn't play nicely with subtyping. Best not to be too hasty with this change, IMO.

axel22 replied:
I agree this needs to be worked out. However, we currently have a request to use custom hashing on TrieMaps. I think we should at least enable the TrieMaps to take custom Hashing objects for now.
Given that we don't do this with the rest of the collections, it should not be hard to switch to a different strategy once we figure it out.

Commit:
Remove Equality in favour of Equiv.

Make Equiv serializable.

axel22 commented:
Commit range updated.

Review by @retronym

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/206/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/206/

jsuereth commented on src/library/scala/collection/concurrent/TrieMap.scala:
```diff
@@ -807,7 +823,16 @@ extends scala.collection.concurrent.Map[K, V]
     if (!RDCSS_ROOT(r, r.gcasRead(this), INode.newRootNode[K, V])) clear()
   }
+  
+  @inline private def mangle(hc: Int): Int = {
```
Just curious, is it still a good idea to mangle hash codes?  If I can provide my own implementation, that's something i'd have to keep in mind when designing the new hash....

axel22 replied:
You have a point there, it would be good to get rid of this.

However, the opposing constraint is that when there is no custom Hashing object, some objects produce poor hashcodes for typical object ranges - for example (n: Int).hashCode == n.
The problem with this is that the upper bits of the hash code are usually always 0. In both concurrent and immutable tries this can cause a slowdown due to several collision levels. In parallel hash tables it causes a similar problem with parallel hash table construction.

So, ideally, we want to mangle the bits when a default is provided, and otherwise not mangle them and hope the user knows what he's doing.

jsuereth replied:
What if we figure out if a default is provided?   We can control the default typeclass, so let's give it a real class name and throw an isInstanceOf before mangling?   Either that or we can wrap the typeclass on the way _in_ to the hash map if it's an instance of default....

I agree, keeping the mangling is good for the default.

axel22 replied:
To avoid `isInstanceOf` on every `computeHash`, I could replace the `hashingobj` with a custom internal `Hashing` which does the mangling if it's a class returned by `Hashing.default`, yes.

jsuereth replied:
+1 ...

jsuereth commented on src/library/scala/math/Hashing.scala:
```diff
@@ -0,0 +1,40 @@
+}
+
+
+object Hashing {
```
Note:  `MurmurHash`  exists in `scala.util`. 

(a) Should we provide a default `MurmurHash` instance?  It's an instance of `T => Int`
(b) Should we adapt `MurmurHash` to extends `Hashing`?
(c) Given that we have two classes, should they both be in the same location?  `scala.math.hash`? `scala.util.hash`? `scala.math`? `scala.util`?

We don't have to move anything round in this CL, but we should make a decision on where we want hashing utilities.

axel22 replied:
I've put the `Hashing` in `math` because I saw `Ordering` there. But this might not have been the best decision - `scala.util` may be a better option.

jsuereth replied:
I could go either way, but I think we should pick one.  I'm fine with either `math` or `util`.  I prefer _NOT_ having any packages called `util` because they become an amalgamation of junk.   How bout we keep them either in `scala.math` or move both to `scala.util.hashing` ?

retronym replied:
Another option: don't integrate `Hash` into the type class hierarchy, but just add the `areEqual` method to it directly.

axel22 replied:
I'd put both `Hashing` and `MurmurHash` in `scala.util.hashing`, perhaps, and put deprecated aliases in `scala.util` for `MurmurHash`.

As for the type class hierarchy - currently `Hashing` is not extending `Equiv`. If such a thing is problematic, I do not think we have to do it yet.
The only current use-case for `Hashing` are `TrieMap`s and the `areEqual` method is not necessary there. Do you think we should have it on `Hashing` otherwise?

axel22 replied:
I created the `scala.util.hashing` package, but did not yet move `MurmurHash` there. We can do this in a separate commit.
Btw, `MurmurHash` is deprecated in favour of `MurmurHash3` - which does not inherit anything yet. Perhaps it should inherit `Hashing`.

jsuereth replied:
Yeah, let's do that in a separate commit.  What you have now is exciting and a good first step.

jsuereth commented on src/library/scala/math/Hashing.scala:
```diff
@@ -0,0 +1,40 @@
+
+object Hashing {
+  
+  implicit def default[T] = new Hashing[T] {
```
Note:  I wish we could use implicit class here...  That would make it simpler to do an isInstanceOf later....

axel22 replied:
Yep, would be cool.

Commit:
Fix `Hashing`.

Move `Hashing` to `scala.util.hashing`.
Adde `Hashing.Default` to `Hashing` companion object.
Change `TrieMap` to autoswitch from `Hashing.Default` to `TrieMap.MangledHashing`.

axel22 commented:
Do we close and re-pull here?

jsuereth commented:
Did you run a build locally?   I'm fine merging what's there now.  If you spawn Lukas's build bot and link the result, that's good enough for me too.

Otherwise, you would have to close and re-pull to get the bot to run again.

axel22 commented:
## Maybe i should close and repull - i didn't rebuild after the 3rd commit.

Aleksandar Prokopec
LAMP, IC, EPFL

Sent from my Android phone with K-9 Mail. Please excuse my brevity.

Josh Suereth reply@reply.github.com wrote:

Did you run a build locally? I'm fine merging what's there now. If you spawn Lukas's build bot and link the result, that's good enough for me too.

Otherwise, you would have to close and re-pull to get the bot to run again.

---

Reply to this email directly or view it on GitHub:
https://github.com/scala/scala/pull/652#issuecomment-6069107
